tornado/ioloop.py

```diff
@@ -927,6 +897,53 @@ def add_callback_from_signal(self, callback, *args, **kwargs):
                 self._callbacks.append(functools.partial(
                     stack_context.wrap(callback), *args, **kwargs))
 
+    def _handle_waker(self, fd, events):
+        try:
+            self._waker.consume()
+        except:
+            # If consume errors out, we want to reset our waker in case it is
+            # left in a bad state.
+            waker_writer_fileno = self._waker.write_fileno()
+            self.remove_handler(self._waker.fileno())
+            self._waker.close()
```

Comment:
Actually, this is unsafe for similar reasons even on windows: another thread may be in the process of calling add_callback and therefore waker.wake(). The switch needs to be done while holding `_callback_lock`.